File: apps/web/src/modules/english-lookup/normalize.ts
const TOKEN_RE = /[A-Za-z]+(?:[-'][A-Za-z]+)*/g
const MAX_LOOKUP_CHARS = 1000

/** Preserve sentence punctuation for translation while normalizing whitespace. */
export function normalizeLookupQuery(raw: string): string {
  return String(raw || '')
    .trim()
    .replace(/’/g, "'")
    .replace(/[–—]/g, '-')
    .replace(/\s+/g, ' ')
}

export function countLookupWords(normalized: string): number {
  if (!normalized) return 0
  return normalized.match(TOKEN_RE)?.length ?? 0
}

export function isValidLookupQuery(normalized: string): boolean {
  return normalized.length > 0 && normalized.length <= MAX_LOOKUP_CHARS && countLookupWords(normalized) > 0
}

export function preferredAudioUrl(audio: {
  us?: string | null
  uk?: string | null
} | null | undefined): string | null {
  if (!audio) return null
  return audio.us || audio.uk || null
}

export function lookupVoiceUrl(query: string, accent: 'us' | 'uk' = 'us'): string {
  return `/api/v1/english-lookup/voice?q=${encodeURIComponent(query)}&accent=${accent}`
}

export function lookupVoicePair(query: string): { us: string; uk: string } {
  return { us: lookupVoiceUrl(query, 'us'), uk: lookupVoiceUrl(query, 'uk') }
}

export function proxiedLookupAudioUrl(src: string | null | undefined): string | null {
  if (!src) return null
  const trimmed = src.trim()
  if (!trimmed) return null
  if (
    trimmed.startsWith('/api/v1/english-lookup/audio') ||
    trimmed.startsWith('/api/v1/english-lookup/voice') ||
    trimmed.startsWith('blob:') ||
    trimmed.startsWith('data:')
  ) {
    return trimmed
  }
  return `/api/v1/english-lookup/audio?url=${encodeURIComponent(trimmed)}`
}

export function mergeLookupAudio(
  current: { us: string | null; uk: string | null } | null | undefined,
  next?: { us: string | null; uk: string | null } | null,
): { us: string | null; uk: string | null } {
  return {
    us: current?.us || next?.us || null,
    uk: current?.uk || next?.uk || null,
  }
}

File: apps/api/src/memory_anki/modules/english_lookup/application/normalize.py
"""Query normalization for Saladict-style 1–5 word English selection."""

from __future__ import annotations

import re

# Hyphenated compounds count as one token (mother-in-law).
_TOKEN_RE = re.compile(r"[A-Za-z]+(?:[-'][A-Za-z]+)*")
_WHITESPACE_RE = re.compile(r"\s+")
_MAX_WORDS = 5
_MAX_TRANSLATION_CHARS = 1000


def normalize_lookup_query(raw: str) -> str:
    """Collapse whitespace; keep letters, hyphens, apostrophes inside tokens."""
    text = str(raw or "").strip()
    if not text:
        return ""
    text = text.replace("’", "'").replace("–", "-").replace("—", "-")
    text = _WHITESPACE_RE.sub(" ", text)
    tokens = _TOKEN_RE.findall(text)
    if not tokens:
        return ""
    return " ".join(tokens)


def count_lookup_words(normalized: str) -> int:
    if not normalized:
        return 0
    return len(normalized.split(" "))


def validate_lookup_query(raw: str) -> tuple[str, int]:
    """Return (normalized, word_count) or raise-ready empty.

    Callers should reject empty or word_count > 5.
    """
    normalized = normalize_lookup_query(raw)
    word_count = count_lookup_words(normalized)
    return normalized, word_count


def is_valid_lookup_query(normalized: str, word_count: int) -> bool:
    return bool(normalized) and 1 <= word_count <= _MAX_WORDS


def normalize_translation_query(raw: str) -> str:
    """Keep sentence punctuation while normalizing whitespace for translation."""
    return _WHITESPACE_RE.sub(" ", str(raw or "").strip())


def is_valid_translation_query(text: str) -> bool:
    return bool(text) and len(text) <= _MAX_TRANSLATION_CHARS


def encode_cambridge_path(text: str) -> str:
    """Saladict: spaces → hyphens, then URI-encode."""
    from urllib.parse import quote

    path = _WHITESPACE_RE.sub("-", text.strip())
    return quote(path, safe="-")


def absolute_url(host: str, link: str | None) -> str:
    if not link:
        return ""
    link = str(link).strip()
    if not link:
        return ""
    host = host.rstrip("/")
    protocol = "https:" if host.startswith("https") else "http:"
    if re.match(r"^[a-zA-Z][a-zA-Z0-9+.-]*:", link):
        return link
    if link.startswith("//"):
        return protocol + link
    if link.startswith("/"):
        return host + link
    return host + "/" + link.lstrip("./")
